IKKBZ Join ordering

Goblin80 opened the pull request:
Finds optimal left-deep tree for an acyclic graph in polynomial time and a necessary prerequisite for search-space linearization.

~~the cost function has a slight implementation mistake due to a misunderstanding on my behalf, which subsequently has a slight effect on the relation rank. i will fix that ASAP.~~

Goblin80 committed:
JoinTree Imp.

Goblin80 committed:
IKKBZ Imp.

Goblin80 committed:
IKKBZ sanity tests

Goblin80 committed:
.clang-format

joka921 commented on src/engine/JoinTree.hpp:
```diff
@@ -0,0 +1,510 @@
+  CHILD,
+
+};
+
```
Hi,
Some initial comments on the code design:

* please split this up into a .h and a .cpp file (or alternatively, if templating this into a similar *.h and *Impl.h file

* We should (for understandability) split the parts of the graph management, and the actual values. 
  So you need  a generic graph that has a generic payload for the nodes (relation names, cardinalities) as well as for the edges (here we will have the selectivities for the IKKBZ).

*

joka921 replied:
* and for your precedence graph building, you should build a new tree, the in-place mutation is not easy to understand.

* so in general, it would be best to use a generic Graph interface that doesn't know anything about IKKBZ and then build IKKBZ on top of this

Goblin80 committed:
split this up to a .h and .cpp

Goblin80 committed:
std::span

Goblin80 committed:
allow discard add_relation

Goblin80 committed:
.clang-format

joka921 commented on src/engine/joinOrdering/QueryGraph.cpp:
```diff
@@ -0,0 +1,412 @@
+  //    hist[n].push_back(b);
+
+  // to be able to apply the inverse operation (QueryGraph::uncombine)
+  // we keep track of the combined relation in the `hist` map
```
As I understand it, the rank function of the combined nodes are wrong, because the cost function is wrong.
I think we should make a plan on how to split this up into a simpler design that is more easily testable.

joka921 commented on src/engine/joinOrdering/IKKBZ.h:
```diff
@@ -0,0 +1,119 @@
+// Mahmoud Khalaf (2024-, khalaf@cs.uni-freiburg.de)
+
+#pragma once
+
```
Okay, to improve this design I would suggest the following first step:

1. Get the combining of nodes right using the follwing design:

```
class Relation {
 // ID, selectivity, cardinality
};

// created by the NormalizeStep.
class CombinedRelations {
  std::vector<Relation>; // contained relations.
  double rank() ; // compute the rank.
}

// Called for two consecutive elements in a chain where the rank is out of order
CombinedRelations combine(const CombinedRelations& first, const CombinedRelations& second);


// Run IKKBZ-Normalize on a single chain until it is fully normalized.
void IkkbzNormalize(std::vector<CombinedRelations>& chain);

// Givien the above structure, this is now simple, just merge by rank.
std::vector<CombinedRelations> IkkbzMerge(std::vector<std::vector<CombinedRelations>>);
```


Maybe start this design in a separate PR, and then star with unit tests for those rather small functions.

Then in a much simpler IKKBZ graph, you can incorporate those structures, but maybe let me know first, 
so that we can talk about the next steps.

joka921 replied:
I think this is the thread that is worth pursuing for now so we can build up a correct and maintainable implementation of IKKBZ.

joka921 reviewed (commented):
Okay, to improve this design I would suggest the following first step:

1. Get the combining of nodes right using the follwing design:

```
class Relation {
 // ID, selectivity, cardinality
};

// created by the NormalizeStep.
class CombinedRelations {
  std::vector<Relation>; // contained relations.
  double rank() ; // compute the rank.
}

// Called for two consecutive elements in a chain where the rank is out of order
CombinedRelations combine(const CombinedRelations& first, const CombinedRelations& second);


// Run IKKBZ-Normalize on a single chain until it is fully normalized.
void IkkbzNormalize(std::vector<CombinedRelations>& chain);

// Givien the above structure, this is now simple, just merge by rank.
std::vector<CombinedRelations> IkkbzMerge(std::vector<std::vector<CombinedRelations>>);
```


Maybe start this design in a separate PR, and then star with unit tests for those rather small functions.

Then in a much simpler IKKBZ graph, you can incorporate those structures, but maybe let me know first, 
so that we can talk about the next steps.

Goblin80 committed:
move out rank and test it separately

Goblin80 committed:
make linker happy

Goblin80 committed:
[skip ci] rank table with CostASI

Goblin80 committed:
cleanup costfn

Goblin80 committed:
link util

Goblin80 committed:
link engine

Goblin80 committed:
[skip ci] argmin of all relations

Goblin80 committed:
[skip ci] ASI comments

Goblin80 committed:
[skip ci] IKKBZ par_unseq

Goblin80 committed:
[skip ci] sprinkle const

Goblin80 committed:
[skip ci] iter relay

Goblin80 committed:
[skip ci] partial_sort

Goblin80 committed:
C span seq

Goblin80 committed:
.clang-format

Goblin80 committed:
gcc11 compatible

Goblin80 committed:
linker stuff

Goblin80 committed:
rm CostASITest.cpp

Goblin80 committed:
include .cpp

Goblin80 committed:
mac being annoying per uge

Goblin80 committed:
Merge branch 'ad-freiburg:master' into joinorder-ikkbz

Goblin80 committed:
decouple cost fn and memorize rank

rm CostASI

decouple cost function

connection weight

GOO draft

memorize rank

Goblin80 committed:
ty abseil

Goblin80 committed:
unpack paired rxs

pre pair hist

subchain root exclude

hist of pairs

mem C, T and rank

unpack tests

Goblin80 committed:
inv default undirected

Goblin80 committed:
init JoinTree

Goblin80 committed:
auto?

Goblin80 committed:
JoinNode, JoinTree

Goblin80 committed:
Cout

Goblin80 committed:
fixup! auto?

Goblin80 committed:
init LinearizedDP

Goblin80 committed:
headers

Goblin80 committed:
TreeCostTest

Goblin80 committed:
Merge branch 'ad-freiburg:master' into joinorder-ikkbz

Goblin80 committed:
fmt

Goblin80 committed:
fixup! fmt

Goblin80 committed:
fix expanded macro

Goblin80 committed:
codecov

Goblin80 committed:
unused

Goblin80 committed:
empty JoinNode

Goblin80 committed:
fixup! empty JoinNode

Goblin80 committed:
fixup! unused

Goblin80 committed:
fixup! codecov

Goblin80 committed:
codeconv stuck
